### PR TITLE
bpo-35091: Objects/listobject.c: Don't rely on signed int overflow in…

### PR DESCRIPTION
**Pull request by izbyshev**
… gallop functions

Signed integer overflow is undefined behavior. Use unsigned types instead.

https://bugs.python.org/issue35091

### CODE REVIEW

File: Objects/listobject.c [hunk 1]
```diff
@@ -1354,62 +1354,57 @@ gallop_left(MergeState *ms, PyObject *key, PyObject **a, Py_ssize_t n, Py_ssize_
...
     }
     a -= hint;
 
-    assert(-1 <= lastofs && lastofs < ofs && ofs <= n);
-    /* Now a[lastofs] < key <= a[ofs], so key belongs somewhere to the
-     * right of lastofs but no farther right than ofs.  Do a binary
+    assert(lastofs <= ofs && ofs <= (size_t)n);
+    /* Now a[lastofs-1] < key <= a[ofs], so key belongs somewhere to the
+     * right of lastofs-1 but no farther right than ofs.  Do a binary
      * search, with invariant a[lastofs-1] < key <= a[ofs].
```
Comment (serhiy-storchaka) [inline Objects/listobject.c:1396]:
```suggestion
      * search, with invariant a[lastofs] < key <= a[ofs].
```

File: Objects/listobject.c [hunk 2]
```diff
@@ -1445,62 +1440,57 @@ gallop_right(MergeState *ms, PyObject *key, PyObject **a, Py_ssize_t n, Py_ssize
...
     }
     a -= hint;
 
-    assert(-1 <= lastofs && lastofs < ofs && ofs <= n);
-    /* Now a[lastofs] <= key < a[ofs], so key belongs somewhere to the
-     * right of lastofs but no farther right than ofs.  Do a binary
+    assert(lastofs <= ofs && ofs <= (size_t)n);
+    /* Now a[lastofs-1] <= key < a[ofs], so key belongs somewhere to the
+     * right of lastofs-1 but no farther right than ofs.  Do a binary
      * search, with invariant a[lastofs-1] <= key < a[ofs].
```
Comment (serhiy-storchaka) [inline Objects/listobject.c:1482]:
```suggestion
      * search, with invariant a[lastofs] <= key < a[ofs].
```

File: Objects/listobject.c (hunk 1 above)
Comment (izbyshev) [reply Objects/listobject.c:1396]: Unlike the first part of the comment, the last line seems to refer to the invariant holding after `++lastofs`, which I moved higher. Therefore, no correction is needed.

**Comment by csabella** [comment]
Merged GH-10202 as the PR approved by @tim-one, so I'll close this one.

**Pull request closed without merging**